Accept nested functions in Dangerous Query Methods

siegfault opened the pull request:
### Summary

I think there’s an opportunity to reduce additional false positives for
Dangerous Query Method deprecations/errors. Similar to https://github.com/rails/rails/pull/36448, it seems
reasonable to allow functions that accept other functions (e.g.
`length(trim(title))`).

### Other Information

Mailing list thread: https://discuss.rubyonrails.org/t/feature-proposal-accept-nested-functions-w-r-t-dangerous-query-methods/78650

*Background*
* PR accepting non-nested functions: https://github.com/rails/rails/pull/36448
* Deep background on deprecation and false positives: https://github.com/rails/rails/issues/32995
* Constants: `COLUMN_NAME` for the first and `COLUMN_NAME_WITH_ORDER` for both

siegfault committed:
Accept nested functions in Dangerous Query Methods

Mailing list thread: https://discuss.rubyonrails.org/t/feature-proposal-accept-nested-functions-w-r-t-dangerous-query-methods/78650

*Summary*
I think there’s an opportunity to reduce additional false positives for
Dangerous Query Method deprecations/errors.

*Nested Functions*
Similar to https://github.com/rails/rails/pull/36448, it seems
reasonable to allow functions that accept other functions (e.g.
`length(trim(title))`).

*Background*
* PR accepting non-nested functions: https://github.com/rails/rails/pull/36448
* Deep background on deprecation and false positives: https://github.com/rails/rails/issues/32995
* Constants: `COLUMN_NAME` for the first and `COLUMN_NAME_WITH_ORDER` for both

ghiculescu commented:
Thanks for the PR. Could you add a test that shows some example of what would be accepted now?

I think ideally https://github.com/rails/rails/pull/33330 would handle all these cases, but it doesn't seem to be moving very quickly.

siegfault commented:
Absolutely - thanks for taking a look! Let me know how those tests look and if there are other places I should add tests. https://github.com/rails/rails/pull/36448 did also modify `activerecord/test/cases/associations/eager_test.rb` and `activerecord/test/cases/relations_test.rb`, but I'm not sure whether that was more out of necessity

siegfault commented:
@ghiculescu I went ahead and fixed some merge conflicts, added a test as you mentioned (but let me know if there are others I should add!), and I think this is ready to go. Let me know if there's any other improvements I can make

ghiculescu commented:
That looks really good. I will leave it for review from a core team member.

ghiculescu commented:
Nah, it just needs someone from the core team to review. They are pretty busy (particularly with Railsconf this week) but I'm sure they will get to it!

ghiculescu commented on activerecord/CHANGELOG.md:
```diff
@@ -62,6 +62,10 @@
 
     *Alex Ghiculescu*
 
+*   Allow nested functions as safe SQL string
```
please move this to the top of the file

siegfault commented:
> particularly with Railsconf this week

Oh, right. Unideal time for me to ping on this :laughing: I'll leave this be. Thanks again for all the help!
